Added ajax Username & Email checks

mxkmp opened the pull request:
Pull Request for Issue #9943 
I continued the great work of @framontb  at @icampus
#### Summary of Changes
- Added ajax username and email checks, which compares the entries of the database (only in Frontend)
- Live checks of the equality of the passwords and emails (Frontend)
- added wildcard/blacklist for email domains(Works on Front & Backend)
#### Testing Instructions
- (backend) Allow User Registration in Front End (Users -> Options -> Allow User Registration)
- Configure Username options in backend
- Enter an username which is already in use
- Enter an email which is already in use
- Enter different passwords or emails in the both fields
- Enter a mail from a blacklisted email domain

brianteeman commented on administrator/components/com_cache/views/purge/tmpl/default.php:
```diff
@@ -10,7 +10,7 @@
 ?>
 
-<form action="<?php echo JRoute::_('index.php?option=com_cache&view=purge'); ?>" method="post" name="adminForm" id="adminForm">
+<form action="<?php echo JRoute::_('index.php?option=com_cache'); ?>" method="post" name="adminForm" id="adminForm">
```
what does this change have to to with username and email checks

mxkmp29 replied:
I pulled the Issue 9943 and this was maybe in there. I didn't changed anything in this .php File. Shall I remove this?

brianteeman replied:
@roland-d said he will help you tomorrow to resolve this

brianteeman commented:
@mxkmp29 there are a lot of unrelated changes in this PR

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -34,6 +34,8 @@ COM_USERS_CONFIG_FIELD_CAPTCHA_DESC="Select the captcha plugin that will be used
 COM_USERS_CONFIG_FIELD_CAPTCHA_LABEL="Captcha"
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_DESC="Allow users to change their Login name when editing their profile."
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_LABEL="Change Login Name"
+COM_USERS_CONFIG_FIELD_EMAIL_REGEX_LABEL="Forbidden email endings"
```
Do you mean forbidden domains?

mxkmp29 replied:
yes, sorry I formulate it false.

brianteeman commented:
Closing for now until the branch is fixed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

Commit:
Implement AJAX user changes

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -34,6 +34,8 @@ COM_USERS_CONFIG_FIELD_CAPTCHA_DESC="Select the captcha plugin that will be used
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_DESC="Allow users to change their Login name when editing their profile."
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_LABEL="Change Login Name"
+COM_USERS_CONFIG_FIELD_EMAIL_REGEX_LABEL="Forbidden domains"
+COM_USERS_CONFIG_FIELD_EMAIL_REGEX_DESC="Expression for forbidden email adress endings. Example: Forbid all email addresses from example.com. Input:<br /> example.com<br />For many entries add a new line after every entry. <br />Input: <br />example.com<br />example2.com"
```
Can we simplify this to

List of forbidden domain names. Place each domain name on a new line

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_FIELD_USERACTIVATION_LABEL="New User Account Activation"
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_ADMINACTIVATION="Administrator"
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_SELFACTIVATION="Self"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_LABEL="Characters preset"
```
Labels should be in capital case eg
Characters Preset

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_MAXNUMCHARS_DESC="MAXimum number of characters required for username field. Zero indicates no limit."
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSET_LABEL="CUSTOM character set"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSET_DESC="CUSTOM character set ALLOWED/FORBIDDEN/REGEXP for username field. Select the appropriate behaviour in previous select list. If other than 'CUSTOM' is selected, this text box is not operational. Example allow/forbid: write 'abc' (without quotations) to allow/forbid these characters. For 'CUSTOM REGEXP' case, write a pattern to use inside a preg_match_a ...
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_DISABLECHECK="DISABLE characters check"
```
character not characters

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_IMPORT_FAILED="An error was encountered while importing the configuration: %s."
 COM_USERS_CONFIG_PASSWORD_OPTIONS="Password Options"
 COM_USERS_CONFIG_SAVE_FAILED="An error was encountered while saving the configuration: %s."
+COM_USERS_CONFIG_PASSWORD_OPTIONS="Password options"
```
Capital case please

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_SAVE_FAILED="An error was encountered while saving the configuration: %s."
+COM_USERS_CONFIG_PASSWORD_OPTIONS="Password options"
 COM_USERS_CONFIG_USER_OPTIONS="User Options"
+COM_USERS_CONFIG_USERNAME_OPTIONS="Username options"
```
Capital case please

brianteeman commented on components/com_users/views/registration/view.html.php:
```diff
@@ -109,7 +108,7 @@ protected function prepareDocument()
 
 		$this->document->setTitle($title);
-
+        //this->document
```
??

brianteeman commented on libraries/joomla/form/rule/username.php:
```diff
@@ -57,6 +62,256 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
+		// Get the minimum number of characters
+		$minNumChars = $params->get('minimum_length_username');
+
+		// If is set minNumChars and $usernameLenght does't achieve minimum lenght
```
typo $usernameLength

brianteeman commented on libraries/joomla/form/rule/username.php:
```diff
@@ -57,6 +62,256 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
+		// Get the maximum number of characters
+		$maxNumChars = $params->get('maximum_length_username');
+
+		// If is set maxNumChars and $usernameLenght surpass maximum lenght
```
typo $usernameLength
typo length

brianteeman commented on libraries/joomla/form/rule/username.php:
```diff
@@ -57,6 +62,256 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
+						. ')$/u';
+
+					// @TODO extract common and inheritaded
+					// Si el username proviene de más de un script recomendado (o combinación permitida) lanza error
```
can we have this comment in english please

Commit:
added lost backend option

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -11,214 +11,274 @@
+                return regex.test(value);
+            });
+            //check if username is already in database
+            jQuery('#jform_username').change(function(){
```
it is com_users specific code it should not be here in validate.js, 
and I do not see a reason why you need change the validate.js at all

mxkmp29 replied:
I thought this should be placed there where the validation of the fields is. Where should I place it that I can integrate it perfectly ?

Fedik replied:
you can make separated file and place it under /media/com_users/js/validate-user.js
or in the layout where the form is rendered

brianteeman replied:
can we try to use vanilla js please and not jquery specific if at all possible.

roland-d replied:
@Fedik @dgt41 Can we have some help here on how to turn an AJAX call to vanilla JS and also the .change() functions like ```jQuery('#jform_username').change(function(){``` I don't know how to do that and we need to make sure it supports IE8.

andrepereiradasilva replied:
i have used one here... https://github.com/joomla/joomla-cms/pull/8545/files#diff-5a22fd5c5d43e73f7858a435aa497b18R30
but may be @Fedik @dgt41 have a better way? (add vannila js ajax calls in core.js?)

dgt41 replied:
I think https://github.com/joomla-projects/media-manager-improvement/blob/b60da9fb9aa9a5893852c4e0926191f86a9479a8/media/plg_media-editor_imagefilters/js/caman-init.js
is a good starting point. There is a polyfill for IE8/9 that might needed as well here.
Check this article for references.
ATM I haven't check the current code, will do when I'll get some free time

Fedik replied:
then we need to think about general AJAX method in `core.js`, 
```javascript
Joomla.request = function(options){ ... }
// do request
Joomla.request({
 url: '/some/url',
 onSuccess: function(){alert('works!')},
 onError: function(){alert('not works!')},
})
```
Also we need the events handling.
About IE8, there `change` does not work, but there can try to use `onpropertychange` instead `change`
This require a bit  more coding 😉 

so for now I would leave it with jQuery. Just move the code from `validate.js`.

ps. when I will get some time, later will try to prepare the pull request for Joomla ajax and events.

roland-d replied:
@Fedik I had a Skype chat with @dgt41 today and we came to the exact same conclusion. So I guess we are all on the same page. Also the student has very limited time, so rather have this complete as it is and we can refactor to native JS once we have all our "own" methods in place. Rather I use jQuery now and later move everything to the J! way than everybody copy-pasting possible solutions from all over the net.

Commit:
fixxed errors & typos

brianteeman commented on administrator/components/com_users/config.xml:
```diff
@@ -185,7 +190,77 @@
+			cols="50"
+			filter="string"
+			>
+			</field>
```
Can we add a showon param here so this is only shown when custom is selected - then you can also simplify the language string

eg
showon="allowed_chars_username_preset:1,2,3"

and then the string you can delete "If other than 'CUSTOM' is selected, this text box is not operational."

Commit:
fixed backend options, moved js to a seperate file, added session tokens

andrepereiradasilva commented on components/com_users/controllers/registration.json.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Profile controller class for Users.
+ *
+ * @since  1.6
```
use always  `@since __DEPLOY_VERSION__`

andrepereiradasilva commented on media/system/js/validate-uncompressed.js:
```diff
@@ -11,214 +11,274 @@
+                var name = jQuery(this).val();
+                if(name.length){
+                    jQuery.ajax({
+                        url: 'index.php?option=com_users&task=registration.validate&format=json&username=' + name,
```
you should add the session token (and verify it in the server side) in the ajax calls

dgt41 replied:
@andrepereiradasilva is right here, best approach is to have a data-url in the input where you just inject the url with the right session token. Then, here in the script, you just parse that and use it.

andrepereiradasilva replied:
used that for the sendtestmail.js ajax call

https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_config/view/application/tmpl/default_mail.php#L34

Fedik commented on components/com_users/views/registration/view.html.php:
```diff
@@ -43,6 +42,10 @@ public function display($tpl = null)
 
+		JText::script('COM_USERS_PROFILE_EMAIL2_MESSAGE');
+		JText::script('COM_USERS_FIELD_RESET_PASSWORD1_MESSAGE');
+		JHtml::_('script', 'com_users/validate-user.js', false, true);
```
Please move it to the layout `components/com_users/views/registration/tmpl/default.php`
Avoid  adding a script/style in the view class.

Fedik commented on media/com_users/js/validate-user.js:
```diff
@@ -0,0 +1,64 @@
+window.onload = function() {
```
Pleas add ` "use strict";` and since jQuery already involved use it 😉 
```javascript
!(function($){
   "use strict";
   
   $(document).ready(function(){
     ... here is your code ...
  });
})(jQuery);
```
Thanks! 😉

Commit:
moved script/style to default.php
added strict mode to validate-users.js

Commit:
Merge branch 'staging' into username

Devportobello commented on components/com_users/controllers/registration.json.php:
```diff
@@ -0,0 +1,76 @@
+	public function validate()
+	{
+		// Check for request forgeries.
+		JSession::checkToken('get') or jexit(JText::_('JINVALID_TOKEN'));
```
Dunno if we need to use jexit() in a json response context, in the case of user afk for a while the csrf will expire and JS .done() not gonna reached

roland-d replied:
@Devportobello What do you suggest we use then if the token is invalid?

Devportobello replied:
@roland-d what about:
```php
$checkToken = JSession::checkToken('get');
if($checkToken === false)
{
	echo new JResponseJson(null,JText::_('JINVALID_TOKEN'),true);
	return false;
}
```

roland-d replied:
@Devportobello Sounds good to me. @mxkmp29 Can you update the code accordingly? Just make sure it follows code-style ;)

Devportobello replied:
@roland-d just dont know how that [case](https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/session/session.php#L321-L322) react on an ajax context :/

roland-d replied:
@Devportobello If you get to that point it returns true and the AJAX request will be performed because there is no failure. Since we are only validating here, I think it should be OK.

Devportobello commented on libraries/joomla/form/rule/username.php:
```diff
@@ -57,6 +62,257 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
+		// If is set minNumChars and $usernameLength does't achieve minimum lenght
+		if (($minNumChars) && ($usernameLength < $minNumChars))
+		{
+			$app->enqueueMessage(JText::sprintf('COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_REQUIRED', $minNumChars, $usernameLength), 'warning');
```
now the rule is depending of the com_users language !

roland-d replied:
@Devportobello Good point, we should load the com_users language file in that case because more strings are used later on in the code as well.

mbabker replied:
Except libraries code should NOT be dependent on a component for anything.  And the fact there are already these types of dependencies doesn't make it right to add more.

Devportobello replied:
@mbabker agreed, but i notice that for B/C

mbabker replied:
It's not just a matter of moving files.  For language strings, they should be added to the `lib_joomla` file.  For component parameters, since this is a libraries level field it should in some way be configurable; right now this is hardcoded explicitly to com_users making it unusable for any other user management system.  Meaning either this field should be moved to com_users or the field be refactored in a way that makes it suitable for multiple components (a simple extension parameter in the XML schema and a default value in the PHP code of com_users would suffice for this part of things).

mbabker replied:
Just for reference, one example where we moved a helper from a component to the libraries to allow it to be re-used across components can be found with https://github.com/joomla/joomla-cms/commit/af994a14c0978d7e8581c6ea9116d211733c8455

Not exactly the same as making a form field extensible, but kind of gives an idea of a way to approach things.

roland-d replied:
Yeah, good points Michael, we will have a look at that tomorrow. Most of the things are new as this is taken from another PR. None of these comments were made there so they were not taken into account.

Devportobello replied:
@mbabker well i understood better now

Devportobello commented on media/com_users/js/validate-user.js:
```diff
@@ -0,0 +1,69 @@
+                    type: 'POST',
+                    url:  ajaxUrl + '&email=' + mail
+                }).done(function (data) {
+                    if (data.success) {
```
if data.success == true there is no error ? am i missing something ?

mxkmp29 replied:
In this case is the success that there is an equal name, if there is no equal name the request failed.

Commit:
improved sessiontoken handling

Commit:
improved language handling for third party context

Devportobello commented on libraries/joomla/form/rule/username.php:
```diff
@@ -35,6 +36,19 @@ class JFormRuleUsername extends JFormRule
+		// Load language files
+		if ($app->isSite())
+		{
+			JFactory::getLanguage()->load('com_users', JPATH_SITE, null, true);
```
i think L44 to L51 could be replaced by: JFactory::getLanguage()->load('com_users');

mxkmp29 replied:
thanks, i've fixed it .

Commit:
bugfixes

mbabker commented on libraries/joomla/form/rule/username.php:
```diff
@@ -57,6 +62,257 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
+
+				case 7:
+					// EMAIL
+					jimport('joomla.mail.helper');
```
JMailHelper is autoloaded, no need to import it.

C-Lodder commented on media/com_users/js/validate-user.js:
```diff
@@ -0,0 +1,80 @@
+                    type: 'POST',
+                    url: ajaxUrl + '&username=' + name
+                }).done(function(data){
+                    console.log(data);
```
You can remove the `console.log()`

Commit:
removed redundant stuff

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 53e1a77c8a207523b43a61c1ee2ea08e31dd7e2d<br /><br />Before this PR there is ONE tab called password options
After this PR there are TWO tabs<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

brianteeman commented on administrator/components/com_users/config.xml:
```diff
@@ -140,7 +140,12 @@
-
+	</fieldset>
+	
+	<fieldset
```
This should not be here. This has created a second Password Options tab

Commit:
removed dublicated password tab

Commit:
removed depency  on com_users
bugfixes on regexpression

Schmidie64 commented:
I have tested this item :white_check_mark: successfully on 809f68aa42d3d894e31396b931ebe897e5c20faf<br /><br />@icampus
I have tested this Pull Request Successfully. I installed the Pull Request and followed the Test Instructions.
All domains i set on the Blacklist were blocked.
Everything work well.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 809f68aa42d3d894e31396b931ebe897e5c20faf<br /><br />The default setting for the number of characters in a password is 0 which means there is no minimum or maximum characters
BUT
Try to save a username with just one character and you will get the following error
Save failed with the following error: Please enter a valid username. No space at beginning or end, at least 2 characters and must not contain the following characters: < > \ " ' % ; ( ) &.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

brianteeman commented:
Entering an email from the blocked list gives a different error in the admin and the frontend

#### admin
> Warning
Invalid field: Email

#### frontend
> Warning
The email address you entered is already in use or invalid. Please enter another email address.

It would be preferable if they were the same and then we can work on some more appropriate text for the message

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

brianteeman commented:
####AJAX
Maybe its me misunderstanding but i am not seeing any ajax validation. Can you please record a small screen capture of what I am supposed to be seeing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

mxkmp29 commented:
![username](https://cloud.githubusercontent.com/assets/9352415/17618318/41d0130e-607f-11e6-9e08-fe52f11a14a3.PNG)
If a username is entered and the field is leaved the ajax check is exectuted.
Thanks for the testing I'll fix the issue with the  username length and the backend message.

mxkmp29 commented:
sorry, I clicked the wrong button

brianteeman commented:
Ah I see now that the AJAX stuff only works on the frontend - it doesnt work on the admin at all

roland-d commented:
@brianteeman The reason you see 2 different messages is because the back-end form uses Javascript validation and the front-end does not. The Javascript validation has no option for using language strings at this moment. It simply reads the invalid field label and concatenates that to the **Invalid field: ** message.

To make this consistent Max would have to rewrite the whole back-end form validation. Even though I agree the messages should be the same, this may be out of scope for this pull request and would warrant another pull request.

In general I believe we should have a more flexible and consistent validation system for forms without duplicating checks.

Your thoughts?

Commit:
min username length adjusted

brianteeman commented:
yes I see now that the ajax stuff is not present on the admin side at all.
- yes it probably is out of scope

On 12 August 2016 at 11:01, RolandD <notifications@github.com> wrote:

> @brianteeman <https://github.com/brianteeman> The reason you see 2
> different messages is because the back-end form uses Javascript validation
> and the front-end does not. The Javascript validation has no option for
> using language strings at this moment. It simply reads the invalid field
> label and concatenates that to the **Invalid field: ** message.
>
> To make this consistent Max would have to rewrite the whole back-end form
> validation. Even though I agree the messages should be the same, this may
> be out of scope for this pull request and would warrant another pull
> request.
>
> In general I believe we should have a more flexible and consistent
> validation system for forms without duplicating checks.
>
> Your thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11516#issuecomment-239408632>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8fNngwWlGZnPeG6IevptfpH8Q0Kpks5qfERxgaJpZM4Je_OM>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

roland-d commented:
Thanks for the input, we will keep it to the front-end for this PR. Max already fixed the username length, so it's ready for testing again.

brianteeman commented:
thanks for fixing the username length - that works

On 12 August 2016 at 11:10, RolandD <notifications@github.com> wrote:

> Thanks for the input, we will keep it to the front-end for this PR. Max
> already fixed the username length, so it's ready for testing again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11516#issuecomment-239410404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8dmov8RTArR4-N36STvEosjLB_aeks5qfEaggaJpZM4Je_OM>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

brianteeman commented:
Is there supposed to be an ajax check on the email? <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

ggppdk commented:
This PR needs a "New feature" label ?

mxkmp29 commented:
@brianteeman yes this is supposed to be

brianteeman commented:
> Max already fixed the username length, so it's ready for testing again.

The language string needs updating now as there isnt a Zero indicates no limit

Commit:
languagefile updated on min limit of username length

brianteeman commented:
Its possible to set the maximum number of characters in the username to less than the minimum

brianteeman commented:
Set the minimum number of characters to 3 and you can still create a 2 character username in the admin.
In the front end you are restricted but the error messages are wrong
You get these two
> Warning
The username field should contain a minimum of 3 characters. You only give 2.
The username you entered is not available. Please pick another username.

You dont need the two messages - and I would be frustrated if I then chose a username that was longer than the maximum

>The username should contain a minimum of 3 characters and a maximum of 5. You only gave 2. Please enter another username.

ggppdk commented:
Some comments, 

Using a comparison tool, i see that : **media/system/js/validate-uncompressed.js**
only has fixed indentationand some new lines, i guess it is ok

Also i see that this can be usable in 3rd party code too

A question should the number of requests to:
$ajaxUri = JRoute::_('index.php?option=com_users&task=registration.validate&format=json&' . JSession::getFormToken() . '=1');

be limited to somehow ?
or is it ok to allow unlimited number ?
someone can grab the above url, from the HTML and make unlimited requests ?

roland-d commented:
@brianteeman We looked at this and it actually works as designed at the moment. I agree with you that we should just show one error message. The problem is the following. To be able to make a customized error message based on the settings of the site, we need to pass it the configuration values. This is impossible with the current way Joomla works. By adding the message to the rule which contains these settings we can customize the message.

Now the XML definition contains an element called message which holds the language string to show when there is an error. This is why we get two messages, the customized one and the default one. Now the idea was to remove the default one but that resulted in **Invalid field: Username**. So that is not the way to go either.

The only solution I see for one message is to modify the existing default ones and forget about the customizing of the message. The default one should be very generic in that case.

@ggppdk The requests are limited by the session token which I believe is good enough. Yes, you can grab that URL and make unlimited requests just like you can take the homepage URL and make unlimited requests. I guess I don't see this as an issue.

brianteeman commented:
If it tells me that I have failed the minimum and try again
and then I try again and it tells me the maximum I will not be a happy user

roland-d commented:
The current default message is:
> The username you entered is not available. Please pick another username.

This we can change to:
> The username does not meet the requirements.

It just won't tell you which requirements because we can't specify them.

Commit:
changed default value

ggppdk commented:
@roland-d 
> The requests are limited by the session token which I believe is good enough. Yes, you can grab that URL and make unlimited requests just like you can take the homepage URL and make unlimited requests. I guess I don't see this as an issue.

ok, i see
i was just wondering if this makes much easier, to find a sub-set of usernames and emails that are used in a website,

JSON requests will also run much faster and also it is much easier to check the response
i guess, related plugins at JED will be updated to handle this case too,

roland-d commented:
@ggppdk As for finding out existing usernames/emails that will depend on the message given I guess. You can also write a script that mimics sending the user registration and parse the output of that. Indeed JSON responses are easier but does it make a real difference with todays computing powers?

ggppdk commented:
@roland-d 
indeed you are right, not too much difference,

and adding built-in  brute-force  prevention is off-topic for this PR  (and i know that there are good plugins at JED for this)

roland-d commented:
@ggppdk You have a valid point but we may want to have something generic then in Joomla. Perhaps with webservices ;)

brianteeman commented:
Just wondering why the maximum number of characters for a username is 99 when the field is varchar(150)

Commit:
changed maximum and minimum for usernames

Commit:
changed border because of maximum in the database

mxkmp29 commented:
because the minimum length have to be 3 characters

brianteeman commented:
Maybe I am being thick but why does an absolute minimum of three prevent you from setting a minimum of 15o

roland-d commented on administrator/components/com_users/config.xml:
```diff
@@ -140,52 +140,137 @@
 			</field>
+		<field
+				name="minimum_length"
+				type="integer"
```
You can change this to number for a number input field.

roland-d commented on administrator/components/com_users/config.xml:
```diff
@@ -140,52 +140,137 @@
-			default="4">
-			</field>
+				name="minimum_integers"
+				type="integer"
```
You can change this to number for a number input field.

roland-d commented on administrator/components/com_users/config.xml:
```diff
@@ -140,52 +140,137 @@
-			default="0">
-			</field>
+				name="minimum_symbols"
+				type="integer"
```
You can change this to number for a number input field.

mxkmp29 replied:
If I type manually in a number an press save, the number will be saved if the type is "number". I think that is a big problem, because the limitation can be skipped easily. 
Maybe I'm missing something for the inputfield. 

> type="number"
> max="149"
> min="0"

are the changes.

@roland-d

mxkmp29 replied:
@roland-d 
what is with the numberinput field?

roland-d replied:
@mxkmp29 I was talking about this field https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/form/fields/number.php

Which has a minimum and maximum value as well. That is not usable here?

mxkmp29 replied:
@roland-d I descripted the problem 2 post above. 
The skip can be skipped with typing a number directly in the input box.

roland-d replied:
Hey Max, I have tested this and I see what you mean now. The maximum value is not respected in a number field. Not something I would expect. Let's leave it at integer.

mbabker replied:
Change it to number and add the `validate="number"` attribute to the field.  This will cause it to be processed through `JFormRuleNumber` which validates the min and max attributes.

mxkmp29 commented:
sorry I was a bit confused. Should the maximum 149 because of the 0 it would be 150 Characters or I'm wrong ?

brianteeman commented:
Yes you are confused. You are confusing the value of an integer with the nu!best of characters in a field

Commit:
*removed specific errormessages (only one errormessage)
*edited errormessage
*length fix

roland-d commented:
Since the DB size of the field is 255 characters, we can set the maximum to 255. The minimum is 3 and up as shown in the message. Those are the correct numbers I think.

Commit:
changed maximum length to 255

Commit:
changed type from integer to number

Commit:
Merge branch 'staging' into username

brianteeman commented:
What is the status of this?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

mxkmp29 commented:
its ready to test @brianteeman

Commit:
Merge branch 'staging' into username

Commit:
Merge branch 'staging' into username

JoshuaLewis commented:
The checks seem to work pretty well, however I do suggest making the warning message go away once you pass validation. For example if the name "Josh" was taken and I change it to "JoshLewis", the message above remains which gives the vibe that both are taken.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

JoshuaLewis commented:
I have tested this item :white_check_mark: successfully on 37a7ec8b8419ffb6397568d21c64530aa3663a15<br /><br />The checks seem to work pretty well, however I do suggest making the warning message go away once you pass validation. For example if the name "Josh" was taken and I change it to "JoshLewis", the message above remains which gives the vibe that both are taken.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

framontb commented:
Happy to see this almost done, @mxkmp29 . An impressive work by the whole team !
It was harder than it seemed at first sight.
Perhaps it should be tested specifically the UNICODE Security MECHANISMS IMPLEMENTATION of Highly Restrictive level (libraries/joomla/form/rule/username.php). I did some tests when I implemented it, but it's worthy a review. The Default Identifier Syntax code (UAX31-D1) is commented because I thought it's too restrictive. Test and insight required too.
Regards. Good job !

Commit:
Merge remote-tracking branch 'remotes/upstream/staging' into username

Conflicts:
	administrator/language/en-GB/en-GB.com_users.ini

Commit:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into username

Commit:
removed messages if the validation doesnt fail
Merge remote-tracking branch 'remotes/upstream/staging' into username

Conflicts:
	administrator/language/en-GB/en-GB.com_users.ini

mxkmp29 commented:
@JoshuaLewis  thanks for the feedback, I changed it.

JoshuaLewis commented:
I have tested this item :white_check_mark: successfully on a8f6ae3adbe6d6e1607dc3b16979ac5d92f0600c<br /><br />Awesome, love the new addition! :-D Tested successfully, works great.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/joomla-cms/11516</a>.</sub>

JoshuaLewis commented:
Looking for more testers. @brianteeman @wilsonge Can you test this as well?

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_ADMINACTIVATION="Administrator"
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_SELFACTIVATION="Self"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_LABEL="Character Preset"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_DESC="Specify from the select list if you want disable this check or allow/forbid CUSTOM character filling the CUSTOM text box. You can also select some of the specified character presettings: alphanumeric, latin, email,..."
```
please change to 
Specify from the select list if you want to disable this check or allow/forbid CUSTOM characters.

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSET_LABEL="Custom character set"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSET_DESC="CUSTOM character set ALLOWED/FORBIDDEN/REGEXP for username field. Select the appropriate behaviour in previous select list. If other than 'CUSTOM' is selected, this text box is not operational. Example allow/forbid: write 'abc' (without quotations) to allow/forbid these characters. For 'CUSTOM REGEXP' case, write a pattern to use inside a preg_match_a ...
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_DISABLECHECK="DISABLE character check"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_ALLOWED="CUSTOM is character allowed"
```
This would be better as simply
Allowed characters

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSET_DESC="CUSTOM character set ALLOWED/FORBIDDEN/REGEXP for username field. Select the appropriate behaviour in previous select list. If other than 'CUSTOM' is selected, this text box is not operational. Example allow/forbid: write 'abc' (without quotations) to allow/forbid these characters. For 'CUSTOM REGEXP' case, write a pattern to use inside a preg_match_a ...
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_DISABLECHECK="DISABLE character check"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_ALLOWED="CUSTOM is character allowed"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_FORBIDDEN="CUSTOM is character forbidden"
```
This would be better as simply
Forbidden characters

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_DISABLECHECK="DISABLE character check"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_ALLOWED="CUSTOM is character allowed"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_FORBIDDEN="CUSTOM is character forbidden"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_REGEXP="CUSTOM is a regular expression"
```
This would be better as simply
Regular expression

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_CUSTOM_REGEXP="CUSTOM is a regular expression"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_ALPHANUMERIC="Alphanumeric Only"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_LATIN="Latin Only"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_OPTION_EMAIL="Email valid address"
```
Better as
Valid email address

brianteeman commented:
@mxkmp29 please address the comments in the code by @roland-d @mbabker and myself

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_SELFACTIVATION="Self"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_LABEL="Character Preset"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_DESC="Specify from the select list if you want disable this check or allow/forbid CUSTOM character filling the CUSTOM text box. You can also select some of the specified character presettings: alphanumeric, latin, email,..."
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_LABEL="Min. number of character"
```
Please change to
Min. number of characters

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_LABEL="Character Preset"
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_DESC="Specify from the select list if you want disable this check or allow/forbid CUSTOM character filling the CUSTOM text box. You can also select some of the specified character presettings: alphanumeric, latin, email,..."
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_LABEL="Min. number of character"
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_DESC="Minimum number of character required for username field."
```
Please change to
Minimum number of characters required for the username.

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_CHARSPRESET_DESC="Specify from the select list if you want disable this check or allow/forbid CUSTOM character filling the CUSTOM text box. You can also select some of the specified character presettings: alphanumeric, latin, email,..."
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_LABEL="Min. number of character"
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_DESC="Minimum number of character required for username field."
+COM_USERS_CONFIG_FIELD_USERNAME_MAXNUMCHARS_LABEL="Max. number of character"
```
Please change to
Max. number of characters

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_LABEL="Min. number of character"
+COM_USERS_CONFIG_FIELD_USERNAME_MINNUMCHARS_DESC="Minimum number of character required for username field."
+COM_USERS_CONFIG_FIELD_USERNAME_MAXNUMCHARS_LABEL="Max. number of character"
+COM_USERS_CONFIG_FIELD_USERNAME_MAXNUMCHARS_DESC="Maximum number of character required for username field. Zero indicates no limit."
```
Please change to
Maximum number of characters required for the username. Zero indicates no limit."

brianteeman commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -67,10 +69,28 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_IMPORT_FAILED="An error was encountered while importing the configuration: %s."
 COM_USERS_CONFIG_PASSWORD_OPTIONS="Password Options"
 COM_USERS_CONFIG_SAVE_FAILED="An error was encountered while saving the configuration: %s."
+COM_USERS_CONFIG_PASSWORD_OPTIONS="Password Options"
```
Why are you adding this string. It already exists on line 89

Commit:
corrected some textstrings

mxkmp29 commented:
@brianteeman the comments from @roland-d & @mbabker are corrected. The comment was on the wrong fields in the code. Under "Username Options" the changes are visible and working, which were discussed in this comment.

zero-24 commented on libraries/joomla/form/rule/username.php:
```diff
@@ -57,6 +62,244 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
+						$result = false;
+					}
+
+					/*
```
why do we have commented code here? maybe it should be removed?

mxkmp29 replied:
This commented code ist form the PR before me. I think I can remove this.

Commit:
Merge remote-tracking branch 'remotes/upstream/staging' into username

JoshuaLewis commented:
I'm willing to retest once the comment zero-24 mentioned is either addressed or explained.

JoshuaLewis commented:
I have tested this item :white_check_mark: successfully on 6046009915031509125edbb7c3112175c04ece0e<br /><br />Works as expected. :-)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/tracker/joomla-cms/11516</a>.</sub>

JoshuaLewis commented:
@brianteeman How does it look now (in terms of the requirements you've mentioned)? After the recent change it's working great on my end.

brianteeman commented:
> @brianteeman the comments from @roland-d & @mbabker are corrected. The comment was on the wrong fields in the code. Under "Username Options" the changes are visible and working, which were discussed in this comment.

I dont see the correction that @mbabker says to make in https://github.com/joomla/joomla-cms/pull/11516/files#r75477191 are you saying thats not correct?

brianteeman commented:
Surely it is not correct that I can set the minimum number of characters to be greater than the maximum number of characters
![screen shot 2016-10-06 at 02 32 07](https://issues.joomla.org/uploads/1/d6682af61fa709ad015a8ef110b1ef72.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/tracker/joomla-cms/11516</a>.</sub>

brianteeman commented:
I am testing this on current 3.7 branch and I am getting no ajax live checking at all - although the rules are working - see the movie below

![ajax](https://cloud.githubusercontent.com/assets/1296369/19144306/a6a2ccd0-8ba0-11e6-9988-918ad7dc5b02.gif)

mxkmp29 commented:
@brianteeman 
>#### Summary of Changes
- Added ajax username and email checks, which compares the entries of the database (only in Frontend)

This happens because the ajax calls only checks usernames or emails in the database, not the validation. The validation was implemented by @framontb in the username.php and contains only checks via php. There is no ajax validation, which checks the the rules from the admin-backend.

brianteeman commented:
Seems daft to me to only do ajax on "if exists" and not on "if valid" kind of defeats the object of the code to me

brianteeman commented:
Also if this is the case that it only checks "if exists" and not "if valid" then the error message is incorrect
> The email address you entered is already in use or invalid.

framontb commented:
@brianteeman 
what you suggest is translate all the validation to ajax?

brianteeman commented:
Well I dont see any point in only having half the job as ajax

brianteeman commented:
There is a possible issue that needs to be considered in that with the ajax checking of existing usernames and email addresses i can quite quickly gather a list of all the users on the site<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/tracker/joomla-cms/11516</a>.</sub>

SniperSister commented:
The username and email usage check is a user enumeration attack vector from a security perspective. Yes, we have this check during the current registration process too, but this one can at least be protected using a captcha, that's not possible for the AJAX request.

So, I would suggest to remove the "is username or email" already used check in this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/tracker/joomla-cms/11516</a>.</sub>

conconnl commented:
You can enter and save a lower number of "Max. number of characters" then "Min. number of characters"
This can' be correct. As you can see in my screenshot I can configure to have a username with at least 3 characters but maximum 1 character

![screen shot 2016-12-02 at 06 42 15](https://issues.joomla.org/uploads/1/39044744ac9a6298b735579c727973c4.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11516">issues.joomla.org/tracker/joomla-cms/11516</a>.</sub>

franz-wohlkoenig commented:
@mxkmp29 is this PR for testing?

jeckodevelopment commented:
@mbabker can I ask the JSST decision/opinion about this PR that may include potential security issues?

mbabker commented:
This is not ready for a review, it still has outstanding change requests from several months ago and just through scanning a couple of files there are miscellaneous whitespace changes littered in the PR.

From a security perspective, the concept is "controversial" and this is introducing a "convenience over security" workflow by making it easier to ascertain whether a username exists on a website.  Considering this is triggered by JavaScript based on typing into a field, the implementation needs to ensure that there isn't a way to turn this into a DoS attack on a site.

mxkmp29 commented:
Hey,
Thanks for the big discussion under this PR, but I don't have the time and maybe not the knowledge about joomla and how to fix the security issues to finish this PR. Maybe someone else can continue this work, if some of these features are needed in a future version of joomla. 

best regards
